Add alicloud provider, including resource and examples of alicloud provider

xiaozhu36 opened the pull request:
This is terraform alicloud provider provided by Official AliCloud. The provider is aim at building, changing, and managing alicloud infrastructure safely and efficiently based on terraform powerful management capabilities for infrastructure. At present, it has supported several alicloud resource, such as ECS, VPC, SLB, Security Group and EIP. With the continuous development and improvement to the provider, it will have more perfect function and support more alicloud resource. We hope the PR to integrate alicloud provider into terraform and help consumer to manage alicloud infrastructure. If you have any question, we are glad to get feedback as soon as possible, thanks a lot.

stack72 commented on builtin/providers/alicloud/resource_alicloud_nat_gateway.go:
```diff
@@ -0,0 +1,308 @@
+			},
+
+			"bandwidth_packages": &schema.Schema{
+				Type: schema.TypeList,
```
As I see only 4 `bandwidth_packages` are allowed, we can add:

`MaxItems: 4` to the schema

We can then remove the if len() > 4 check below

stack72 commented on builtin/providers/alicloud/resource_alicloud_nat_gateway.go:
```diff
@@ -0,0 +1,308 @@
+	if v, ok := d.GetOk("name"); ok {
+		name = v.(string)
+	}
+	// } else {
```
Can this be removed or is it a todo?

stack72 commented on builtin/providers/alicloud/resource_alicloud_nat_gateway.go:
```diff
@@ -0,0 +1,308 @@
+	d.SetPartial("spec")
+	d.SetPartial("vpc_id")
+
+	// for i, packageId := range resp.BandwidthPackageIds.BandwidthPackageId {
```
redundant comment?

stack72 commented on builtin/providers/alicloud/resource_alicloud_nat_gateway.go:
```diff
@@ -0,0 +1,308 @@
+	}
+
+	d.SetId(resp.NatGatewayId)
+	d.Partial(true)
```
We don't need d.Partial in Create

stack72 commented on builtin/providers/alicloud/resource_alicloud_nat_gateway.go:
```diff
@@ -0,0 +1,308 @@
+			spec = NatGatewaySmallSpec
+		}
+
+		args := &ModifyNatGatewaySpecArgs{
```
Follow the pattern of a single conn.ModifyNatGatewaySpec and pass the args in

xiaozhu36 commented:
Hi @stack72 

Thanks very much for your help and review advise, and our team have improve existing defects in my provider as per your advise. 

Now, we want to pull a new request and the PR has the following changes:

1. Improved the existing defects and fixed some bugs, including removing d.SetPartial in the Create funcs, improving Updatefunc, adding some attributes for the some parameters and so on.

2. Using `govendor` added the SDK correctly and updated the dependencies. 

3. Ran commands 'make fmt' and 'make test vet', and passed it successfully. 

Please review them and looking forward to a good news. Thanks very much.

Thanks

Guimin He

stack72 commented:
Closed via #11235
